climada_petals/hazard/copernicus_forecast/handler.py

```diff
@@ -0,0 +1,560 @@
+"""
+This file is part of CLIMADA.
+
+Copyright (C) 2017 ETH Zurich, CLIMADA contributors listed in AUTHORS.
+
+CLIMADA is free software: you can redistribute it and/or modify it under the
+terms of the GNU General Public License as published by the Free
+Software Foundation, version 3.
+
+CLIMADA is distributed in the hope that it will be useful, but WITHOUT ANY
+WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A
+PARTICULAR PURPOSE.  See the GNU General Public License for more details.
+
+You should have received a copy of the GNU General Public License along
+with CLIMADA. If not, see <https://www.gnu.org/licenses/>.
+
+---
+
+Module to handle seasonal forecast data from the Copernicus Climate Data Store (CDS)
+in the U-CLIMADAPT project.
+
+TODO: update prerequisites
+Prerequisites:
+1. CDS API client installation:
+   pip install cdsapi
+
+2. CDS account and API key:
+   Register at https://cds-beta.climate.copernicus.eu
+
+3. CDS API configuration:
+   Create a .cdsapirc file in your home directory with your API key and URL.
+   For instructions, visit:
+   https://cds-beta.climate.copernicus.eu/how-to-api#install-the-cds-api-client
+"""
+
+import os
+import logging
+import calendar
+import re
+
+import xarray as xr
+import pandas as pd
+import numpy as np
+import cdsapi
+
+from climada.hazard import Hazard
+from climada.util.coordinates import get_country_geometries
+import climada_petals.hazard.copernicus_forecast.indicator as indicator
+
+
+LOGGER = logging.getLogger(__name__)
+
+class ForecastHandler:
```

Comment:
Why is it called ForecastHandler? Is this not simply a class to download any type of data from the Copernicus data store?

Reply:
Currently, it’s called ForecastHandler because its main function is to download, process, and convert data into indices, which are then used to generate Hazard objects. However, this will evolve as the module is restructured. The naming of the scripts is still very flexible. Valentin and I considered options like 1. descargar.py, 2. forecast_index_downloader.py, and 3. indices.py (I know Lukas isn’t a fan of plurals). But this is entirely open for discussion, so if you have any suggestions, they would be more than welcome!

Reply:
My question was more about the word 'forecast'. Since this could also handle any type of weather data, not only forecasts.

Reply:
For the demo, we are currently using only forecast data because the data architecture is specifically structured for seasonal forecasts, such as those from DWD, UK Met Office, Météo France, CMCC, NCEP, and JMA, which may differ from climate data. Seasonal forecast data is time-sensitive and focuses on predicting short- to medium-term trends (weeks to a maximum of 7 months) based on initialized models. It takes into account specific initial conditions, like the forecast month and hourly resolution, to build indices such as heatwaves or temperature extremes (e.g., TX30, TR).

If we want to calculate indices for other datasets, we need to first study the architecture of that data and then test it. For now, only seasonal forecast data can be handled effectively.

Reply:
The current class only handles forecast data from a specific seasonal forecast. We plan to add a more general class (e.g. "CopernicusDownloader") that simply downloads any copernicus data set (and saves it to netcdf eg). The current forecast handler then could then use this download class and processes the data. 
A general "handler" that calculates indices for both forecast and other data types might be more difficult because the downloaded data set probably have very different structures, for instance the member dimension of forecasts. Does this make sense?

Reply:
Yes, thanks for clarifying.

Reply:
Thanks for the comments! The name fits right now since we handle only forecast data, but we’re open to changing it as the module evolves. As was mentioned, we will discuss adding a general downloader class to support other data types, and if that happens, I will adjust the naming accordingly.